pyscf/adc/radc_ip_cvs.py

```diff
@@ -0,0 +1,1101 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Author: Samragni Banerjee <samragnibanerjee4@gmail.com>
+#         Alexander Sokolov <alexander.y.sokolov@gmail.com>
+#
+
+'''
+Restricted algebraic diagrammatic construction
+'''
+import numpy as np
+import pyscf.ao2mo as ao2mo
+from pyscf import lib
+from pyscf.lib import logger
+from pyscf.adc import radc
+from pyscf.adc import radc_ao2mo
+from pyscf.adc import dfadc
+from pyscf import __config__
+from pyscf import df
+from pyscf import symm
+
+
+def get_imds(adc, eris=None):
+
+    cput0 = (logger.process_clock(), logger.perf_counter())
+    log = logger.Logger(adc.stdout, adc.verbose)
+
+    if adc.method not in ("adc(2)", "adc(2)-x", "adc(3)"):
+        raise NotImplementedError(adc.method)
+
+    method = adc.method
+
+    t1 = adc.t1
+    t2 = adc.t2
+
+    t1_2 = t1[0]
+
+    nocc = adc._nocc
+    ncvs = adc.ncvs
+
+    e_occ = adc.mo_energy[:nocc]
+    e_vir = adc.mo_energy[nocc:]
+
+    e_cvs = adc.mo_energy[:ncvs]
+    idn_cvs = np.identity(ncvs)
+
+    if eris is None:
+        eris = adc.transform_integrals()
+
+    eris_ovvo = eris.ovvo
+
+    # i-j block
+    # Zeroth-order terms
+
+    M_ij = lib.einsum('ij,j->ij', idn_cvs ,e_cvs)
+
+    # Second-order terms
+    t2_1 = t2[0][:]
+    t2_1_coee = t2_1[:ncvs,:,:,:].copy()
+    t2_1_ocee = t2_1[:,:ncvs,:,:].copy()
+    eris_ceeo = eris_ovvo[:ncvs,:,:,:].copy()
+    
+    M_ij += 0.5 * 0.5 *  lib.einsum('ilde,jdel->ij',t2_1_coee, eris_ceeo,optimize=True)
+    M_ij -= 0.5 * 0.5 *  lib.einsum('lide,jdel->ij',t2_1_ocee, eris_ceeo,optimize=True)
+    M_ij -= 0.5 * 0.5 *  lib.einsum('ilde,jedl->ij',t2_1_coee, eris_ceeo,optimize=True)
+    M_ij += 0.5 * 0.5 *  lib.einsum('lide,jedl->ij',t2_1_ocee, eris_ceeo,optimize=True)
+    M_ij += 0.5 * lib.einsum('ilde,jdel->ij',t2_1_coee, eris_ceeo,optimize=True)
+
+    M_ij += 0.5 * 0.5 *  lib.einsum('jlde,idel->ij',t2_1_coee, eris_ceeo,optimize=True)
+    M_ij -= 0.5 * 0.5 *  lib.einsum('ljde,idel->ij',t2_1_ocee, eris_ceeo,optimize=True)
+    M_ij -= 0.5 * 0.5 *  lib.einsum('jlde,iedl->ij',t2_1_coee, eris_ceeo,optimize=True)
+    M_ij += 0.5 * 0.5 *  lib.einsum('ljde,iedl->ij',t2_1_ocee, eris_ceeo,optimize=True)
+    M_ij += 0.5 * lib.einsum('jlde,idel->ij',t2_1_coee, eris_ceeo,optimize=True)
+
+    del t2_1
+
+    cput0 = log.timer_debug1("Completed M_ij second-order terms ADC(2) calculation", *cput0)
+    # Third-order terms
+
+    if (method == "adc(3)"):
+
+        eris_oovv = eris.oovv
+        eris_ovoo = eris.ovoo
+        eris_oooo = eris.oooo
+
+        eris_oecc = eris_ovoo[:,:,:ncvs,:ncvs].copy()
```

Comment:
Check for redundant integrals

Reply:
Will address this when optimizing the code. Currently, eris_ceee integrals aren't properly calculated, but rather eris_oeee is sliced to get eris_ceee